InteractiveBlockElements in DCR

gtrufitt opened the pull request:
## What does this change?

- Adds interactive block element to render a link on the server
- Adds a portal for interactive block elements, and loads this with loadable components to split out the component.
- Adds curl AMD loader when we have an interactive block element on the page. It is pasted directly in, the loader has not been updated since 2016.
- Loads the script from the `scriptUrl` via `window.require` which is accessible due to loading the curl loader

Depends on https://github.com/guardian/dotcom-rendering/pull/2616 and https://github.com/guardian/frontend/pull/23547 and (elementId) https://github.com/guardian/dotcom-rendering/pull/2615

## Screenshots

![image](https://user-images.githubusercontent.com/638051/109181158-86262500-7783-11eb-89c0-308257f87fd8.png)

![image](https://user-images.githubusercontent.com/638051/109181852-4ca1e980-7784-11eb-9eb4-7ae4a84fc5a3.png)

## Follow up work

As mentioned in the [documentation](https://github.com/guardian/dotcom-rendering/pull/2608/files#diff-7535f97e603cacda22b68f5923c7440a074dfa33e16324bfbabdd858ed59f75dR10-R65), there is work required to remove boot.js, which is used for 99% of articles with interactive embed elements. This means we don't need to use the curl loader for the majority of pageviews with an interactive embed. PR to follow.

gtrufitt commented on src/web/server/htmlTemplate.ts:
```diff
@@ -214,6 +214,15 @@ export const htmlTemplate = ({
                     })(window, document);
                 </script>
 
+				<script>
+					window.curlConfig = {
+						baseUrl: '${CDN}assets',
+						apiName: 'require'
+					};
+					window.curl = window.curlConfig;
+				</script>
+
```
This is the contract for the curl loader

gtrufitt commented on src/web/server/document.tsx:
```diff
@@ -173,22 +178,33 @@ export const document = ({ data }: Props): string => {
+			},
 			...arrayOfLoadableScriptObjects, // This includes the 'react' entry point
-		].filter(Boolean),
+		].filter(isDefined), // We use the TypeGuard to keep TS happy
```
It wasn't happy with the previous falsy bouncer, had to add a specific Type Guard here for TS to pass

gtrufitt commented on src/web/components/elements/InteractiveBlockComponent.tsx:
```diff
@@ -0,0 +1,104 @@
+		if (scriptUrl) {
+			// We're going to use curl AMD loader to load the script that the
+			// interactive has given us.
+			window.require(
```
Matches up with https://github.com/guardian/frontend/blob/e38da96c59eaaef143b9f89b2318eab9dcf2fc5a/common/app/views/fragments/inlineJSNonBlocking.scala.html#L52

gtrufitt commented on .prettierignore:
```diff
@@ -19,3 +19,6 @@ node_modules/
 !/scripts/**/*.js
+
+# Ignore specific files
+/**/*/curl-with-js-and-domReady.js
```
This file is already minifed and compiled, we don't want to touch it

buck06191 commented on src/web/components/elements/InteractiveBlockComponent.tsx:
```diff
@@ -0,0 +1,102 @@
+It has not been updated since 2016.
+
+MIGRATION FROM FRONTEND
+- TODO For the standard boot.js, we will re-write the behavior in modern JS to avoid the requirement of an AMD loader
```
Is this still a TODO?

gtrufitt replied:
Yup, will follow up PR.

buck06191 reviewed (approved):
It LGTM, but want to check first on a `TODO` inside some inline documentation? Is this to do as part of this PR, or is it future work?

oliverlloyd commented on cypress/integration/e2e/article.elements.spec.js:
```diff
@@ -117,12 +117,50 @@ describe('Elements', function () {
+			cy.visit(
+				'Article?url=https://www.theguardian.com/us-news/2017/jan/17/donald-trump-america-great-again-northampton-county-pennsylvania',
+			);
+			console.log(getIframeBody());
```
Can remove

oliverlloyd commented on index.d.ts:
```diff
@@ -730,7 +731,8 @@ type IslandType =
 	| 'video-facebook-block-element'
-	| 'vine-block-element';
+	| 'vine-block-element'
+	| 'interactive-block';
```
Heads up: this will choke when elementsId lands but is simple to fix

oliverlloyd commented on src/lib/content.d.ts:
```diff
@@ -215,13 +215,15 @@ interface InteractiveAtomBlockElement extends InteractiveAtomBlockElementBase {
 	scriptUrl?: string;
 	alt?: string;
 	role?: RoleType;
+	interactiveBlockIndex?: number;
```
Note to remove this when merging in elementId

oliverlloyd commented on src/web/lib/ElementRenderer.tsx:
```diff
@@ -335,6 +336,16 @@ export const ElementRenderer = ({
 			);
+		case 'model.dotcomrendering.pageElements.InteractiveBlockElement':
+			return (
+				<Figure role={element.role} id={`interactive-block-${index}`}>
```
Note to remove the id at merge

oliverlloyd commented on src/web/server/document.tsx:
```diff
@@ -173,22 +178,33 @@ export const document = ({ data }: Props): string => {
+			},
 			...arrayOfLoadableScriptObjects, // This includes the 'react' entry point
-		].filter(Boolean),
+		].filter(isDefined), // We use the TypeGuard to keep TS happy
```
If we inline this does that make the code simpler or add a bunch of ts issues?

gtrufitt replied:
I think this is a rare occasion where inlining reduced comprehension because we lose the function name that specifically helps with the comprehension

oliverlloyd reviewed (approved):
Thanks for keeping this understandable when its actually a very complex area. I've made some notes to try and make the merge of elementId easier

buck06191 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [interactiveBlockElement - Parity with curl AMD loader](https://trello.com/c/iVox5AwF/1849-interactiveblockelement-parity-with-curl-amd-loader)

gtrufitt commented:
> Thanks for keeping this understandable when its actually a very complex area. I've made some notes to try and make the merge of elementId easier

👍 Thanks, elementId in and those changes made

oliverlloyd commented on cypress/integration/e2e/article.elements.spec.js:
```diff
@@ -117,12 +117,49 @@ describe('Elements', function () {
+			const getIframeBody = () => {
+				return cy
+					.get(
+						'[data-cypress="interactive-element-LA%20Rams%20dead%20cap%20numbers"] > iframe',
```
Maybe you can simplify these data attributes to be easier to match?

oliverlloyd commented on src/web/components/elements/InteractiveBlockComponent.tsx:
```diff
@@ -0,0 +1,105 @@
+
+	return (
+		<div
+			data-cypress={`interactive-element-${encodeURI(alt || '')}`}
```
This could probably be simpler, maybe just `interactive-element`?

gtrufitt replied:
I need a specific interactive-element on the page, and I don't have an index?

oliverlloyd replied:
The test article only has one interactive element so maybe it just works?

oliverlloyd replied:
Alternatively, we could base64 encode the url so that the selector doesn't have weird characters in it?

gtrufitt replied:
> The test article only has one interactive element so maybe it just works?

There's multiple in one of the test pages.

> Alternatively, we could base64 encode the url so that the selector doesn't have weird characters in it?

Sure, don't mind tbh, think maybe that that is 6-of-one-etc because for a selector for the test environment you either have an encoded uri or encoded base 64... I guess it comes to what's better on PROD? But base64 encoded URL will still be pretty long. Ideally there'd be a string -> short ID algo built in to JS..?

oliverlloyd replied:
🤔 I was only saying all these things becuase I saw your Cypress tests were failing on that selector but they seem to pass now so sure, whatever works

gtrufitt replied:
> I was only saying all these things becuase I saw your Cypress tests were failing on that selector but they seem to pass now so sure, whatever works

Oh that was because I forgot to add `interactiveBlockElement` to `typesThatNeedHydrating` 😆

oliverlloyd commented:
Something off with Storybook, it can'd find `window.require`

buck06191 commented:
> Something off with Storybook, it can'd find `window.require`

Is it to do with it running it all in an `iframe`?

gtrufitt commented:
> Something off with Storybook, it can'd find window.require
> Is it to do with it running it all in an iframe?

Locally storybook could find the static Curl loader file (a plain JS file that gets copied over and served as-is). I added this to `preview-head.html` which storybook uses as the `<head>` but Chromatic's storybook couldn't find it, 404ing on the request.

Hidden in the [Chromatic docs](https://www.chromatic.com/docs/setup#run-chromatic):

> We use the build-storybook script from your package.json by default. If you have customized the storybook script (for example, adding a static directory with -s), check that you've done the same for <script>build-storybook.

We *had* customised the script with a static directory for `storybook` but not `build-storybook`.

I added that here: [`055b49f` (#2608)](https://github.com/guardian/dotcom-rendering/pull/2608/commits/055b49fb93f83880e0dcc3e2ec269b7c36aecf2f)
